src/encoding.rs

```diff
@@ -716,6 +716,44 @@ macro_rules! length_delimited {
 pub mod string {
     use super::*;
 
+    // String::as_mut_vec is unsafe because it doesn't check that the bytes
+    // inserted into the resulting vec are valid UTF-8. We check
+    // after fully extending the buffer in order to ensure this is safe.
+    // If an unwind happens before the check is performed and the result
+    // is committed, the string is truncated to the previously valid length.
+    struct StringGuard<'a> {
+        string: &'a mut String,
```

Comment:
From my (admittedly weak) understanding of the unsafe rules, it's instant UB to create a `str`/`String` containing invalid UTF-8.  Given that, I think this pattern is still unsafe, since it's appending bytes to a `String`, then checking validity.  Given my understanding, I think the only safe way to do what we are trying to do here is accumulate into a `Vec<u8>`, check if it's valid utf-8, and then and only then convert to `String`.

Reply:
As far as I understand, it's not UB until the invalid data are accessed as `str`/`String`; but, theoretically, `.as_bytes()` or even `.as_mut_vec()` might be in violation of this.